src/main/java/org/jitsi/jibri/capture/ffmpeg/executor/impl/AbstractFfmpegExecutor.kt

```diff
@@ -0,0 +1,120 @@
+    private val executor = Executors.newSingleThreadExecutor(NameableThreadFactory("AbstractFfmpegExecutor"))
+    /**
+     * The currently active (if any) Ffmpeg process
+     */
+    var currentFfmpegProc: Process? = null
+    /**
+     * In order to both analyze Ffmpeg's output live and log it to a file, we'll
+     * tee its stdout stream so that both consumers can get all of its output.
+     */
+    var ffmpegOutputTee: Tee? = null
+    /**
+     * We'll use this to monitor the stdout output of the Ffmpeg process
+     */
+    var ffmpegTail: Tail? = null
+
+    init {
+        ffmpegOutputLogger.useParentHandlers = false
+        ffmpegOutputLogger.addHandler(FfmpegFileHandler())
+    }
+
+    /**
+     * Get the shell command to use to launch ffmpeg
+     */
+    protected abstract fun getFfmpegCommand(ffmpegExecutorParams: FfmpegExecutorParams, sink: Sink): String
+
+    override fun launchFfmpeg(ffmpegExecutorParams: FfmpegExecutorParams, sink: Sink) {
+        processBuilder.command(getFfmpegCommand(ffmpegExecutorParams, sink).split(" "))
+        processBuilder.redirectErrorStream(true)
+        logger.info("Running ffmpeg command:\n ${processBuilder.command()}")
+        currentFfmpegProc = processBuilder.start()
+        // Tee ffmpeg's output so that we can analyze its status and log everything
+        ffmpegOutputTee = Tee(currentFfmpegProc!!.inputStream)
+        ffmpegTail = Tail(ffmpegOutputTee!!.addBranch())
+        // Read from a tee branch and log to a file
+        executor.submit {
+            val reader = BufferedReader(InputStreamReader(ffmpegOutputTee!!.addBranch()))
+            while (true) {
+                ffmpegOutputLogger.info(reader.readLine())
+            }
+        }
+
+        logger.debug("Launched ffmpeg, is it alive? ${currentFfmpegProc?.isAlive}")
+    }
+
+    override fun getExitCode(): Int? {
+        if (currentFfmpegProc?.isAlive == true) {
+            return null
+        }
+        return currentFfmpegProc?.exitValue()
+    }
+
+    override fun isHealthy(): Boolean {
+        //TODO: should we only consider 2 sequential instances of not getting a frame=
+        // encoding line "unhealthy"?
+        currentFfmpegProc?.let {
+            val ffmpegOutput = ffmpegTail?.mostRecentLine ?: ""
+            val parsedOutputLine = OutputParser().parse(ffmpegOutput)
+            if (!it.isAlive) {
+                logger.error("Ffmpeg is no longer running, its most recent output line was: $ffmpegOutput")
+                return false
+            }
+
+            if (!parsedOutputLine.containsKey("frame")) {
+                logger.error("Ffmpeg is running but doesn't appear to be encoding.  " +
+                        "Its most recent output line was $ffmpegOutput")
+                return false
+            } else {
+                logger.debug("Ffmpeg appears healthy: $parsedOutputLine")
+                return true
+            }
+        }
+        return false
```

Comment:
Okay, I'm confused on this kotlin syntax. `currentFfmpegProc?.let` is essentially a `if currentFfmpegProc { // do logic }`? (It almost looked to to me you were passing in some function. I don't see you using this .let {} syntax often.)

Reply:
yup, that's it.  main reason for doing it this way is as follows:

if you do
```
if (currentFfmpegProc != null) {
   // logic
}
```
you don't actually get a guarantee that currentFfmpegProc will not be null inside the if, as another thread could have come along and changed it.  so you'd do something like
```
val currentFfmpegProcCopy = currentFfmpegProc // currentFfmpegProcCopy is an immutable type
if (currentFfmpegProcCopy) {
   // logic
}
```
since we copied the reference into an immutable variable, we know it couldn't have been changed to null in between the if check and the block inside.  `let` is basically just shortcut code to do this.